discovery.go

```diff
@@ -103,3 +115,55 @@ func exists(path string) bool {
 	_, err := os.Stat(path)
 	return err == nil || !os.IsNotExist(err)
 }
+
+type commandDescription struct {
+	path     string
+	Name     string              `json:"name"`
+	Summary  string              `json:"summary"`
+	Commands commandDescriptions `json:"commands,omitempty"`
+}
+
+type commandDescriptions []*commandDescription
+
+func (d *commandDescription) ToCommand(discoveredIn, root string, args []string, parent Module, depth, maxDepth int) Command {
+	c := &executableCommand{
+		parent:       parent,
+		discoveredIn: discoveredIn,
+		path:         root,
+		args:         args,
+		name:         d.Name,
+		summary:      d.Summary,
+	}
+
+	if len(d.Commands) > 0 && (maxDepth == -1 || depth < maxDepth) {
+		m := &executableModule{executableCommand: *c}
+		m.cmds = d.Commands.ToCommands(discoveredIn, root, append(args, m.Name()), m, depth+1, maxDepth)
+		return m
+	} else {
+		return c
+	}
+}
+
+func (descriptions commandDescriptions) ToCommands(discoveredIn, root string, args []string, parent Module, depth, maxDepth int) []Command {
+	var cmds []Command
+	for _, d := range descriptions {
+		cmds = append(cmds, d.ToCommand(discoveredIn, root, args, parent, depth, maxDepth))
+	}
+	return cmds
+}
+
+func inspect(executable string) (*commandDescription, error) {
+	cmd := exec.Command(executable, "--inspect")
```

Comment:
Is `--inspect` the right name here? (https://github.com/squareup/squareroot-rfcs/pull/12#discussion_r1548588767)

Its job is to describe (in JSON) the commands a CLI application may respond to

Reply:
This is a good callout.

`--inspect` made sense to me because, in Ruby, the `inspect` method works like Go's `%#v` verb — but I don't think it works well outside of that context.

`--describe` is better.

I wonder if going more explicit is even better. Like:
- `--describe-commands`
- `--describe-subcommands`
- `--list-commands`

Reply:
I think `--describe-commands` works great! Ship it!